### DISCUSSION TITLE
exit command is not working through SSH Operator

### DISCUSSION THREAD

**Discussion by dgedela** [apache/airflow#19312] [General]
### Description

DAG:
jps_process_count_command = """
   set -e;
   ps_count=`jps | grep -v Jps | wc -l`
   if [ $ps_count -ne 0 ]
   then
       echo "ps_count is $ps_count, There might be some processes are running"
       exit 1
   else
       echo "All processed were stopped..!"
       exit 1
   fi
"""

jps_process_count = SSHOperator(
    task_id="jps_process_count",
    command=jps_process_count_command,
    ssh_hook=aws_01_ssh,
    dag=aws_server_process_count_dag
)

Output
==========
airflow.exceptions.AirflowException: SSH operator error: error running cmd: 
   set -e;
   ps_count=`jps | grep -v Jps | wc -l`
   if [ $ps_count -ne 0 ]
   then
       echo "ps_count is $ps_count, There might be some processes are running"
       exit 1
   else
       echo "All processed were stopped..!"
       exit 1
   fi, error: 

Doubt:
For Clarification, whenever ps_count is non-zero then the task should be failed and tell that there were some process are running.

Please let me know if we have any other approach to achieve this requirement.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
Converted it to a discussion (it's definitely not a feature issue). 
I suggest you to clarify what you are asking for if you want to get feedback/help as it's entirely unclear what you are asking for.

**Reply by dgedela** [reply]
I thought that it is feature issue as ssh_operator is not supporting for exit.

Please help me on this.

**Reply by potiuk** [reply]
Could you please describe WHAT ssh_operator is not supporting? You need to describe:
1) what you want to achieve
2) what you did not get and why you think it's an error.

Without textual description what you want - just pasting the code without explanation what you want is useless. I have no idea even what the problem is.

**Reply by dgedela** [reply]
Ohh, Sorry for that @potiuk 

what i want to achieve:
1. I have to execute a script in the remote server like process_count.sh
2. process_count.sh script like below.
           ps_count=jps | grep -v Jps | wc -l
           if [ $ps_count -ne 0 ]
           then
                  echo "ps_count is $ps_count, There might be some processes are running"
                  exit 1
          else
                  echo "All processed were stopped..!"
          fi
3. if the rerun count is non-zero then we should exit (meaning, Airflow task should be failed), otherwise task should be succeeded.

why I'm thinking that this is an error:
1. Airflow ssh_operator unable to interpret the command exit
2. logs as follows.
airflow.exceptions.AirflowException: SSH operator error: error running cmd:
set -e;
ps_count=jps | grep -v Jps | wc -l
if [ $ps_count -ne 0 ]
then
echo "ps_count is $ps_count, There might be some processes are running"
exit 1
else
echo "All processed were stopped..!"
exit 1
fi, error:


I hope that im clear now.

**Reply by potiuk** [reply]
No not clear. What do you mean by `Airflow ssh_operator unable to interpret the command exit` ? what did you  to see expect? what error did you see? which part failed? 

Your logs are not telling me what problem you have. Please be clear about it and explain it in words.

**Reply by dgedela** [reply]
Okay ..! I will try to explain clearly

What i want to achieve:

1. I have to execute a script(Process-count.sh) in the remote machine through SSH Operator.
2. And, There is if-else condition in the same script.
3. If the condition satisfies, task is successful otherwise it is failure.

Problem Statement:
1. In My Remote Server, Couple of application processes are running.
2. When We want to restart, All Processes should be stopped before starting.
3. So, Before starting the services, We added one script which checks all processes are stopped.
4. For Point3, We have written a condition, If the Process count is '0' then we are good (Task is success) otherwise, we have to fail the task (To fail - We used 'exit' command)
5. But, Fail Scenario is not working

**Now, The question is how can we fail the task ?**

**what & how we tried (Code is as below):
aws_dev_jps_process_count_dag:**


from airflow import DAG
from datetime import datetime, timedelta
from airflow.operators.dummy_operator import DummyOperator
from airflow.contrib.hooks.ssh_hook import SSHHook
from airflow.contrib.operators.ssh_operator import SSHOperator

default_args = {
	'owner': 'airflow',
	'depends_on_past': False,
	'start_date': datetime(2021, 8, 13),
	'retries': 0
}

aws_dev_process_count_dag = DAG(
    dag_id='aws_dev_process_count_dag',
    default_args=default_args,
    catchup=False,
    schedule_interval='59 05 * * 1-5'
)


aws_dev_01_ssh = SSHHook(ssh_conn_id="aws-dev")

jps_process_count_command = """
   set -e;
   ps_count=`jps | grep -v Jps | wc -l`
   if [ $ps_count -ne 0 ]
   then
       **echo "ps_count is $ps_count, There might be some processes are running, Hence, We are failing it"
       exit 1**
   else
       **echo "All processed were stopped..!"**
   fi
"""

jps_process_count = SSHOperator(
    task_id="jps_process_count",
    command=jps_process_count_command,
    ssh_hook=aws_dev_01_ssh,
    dag=aws_dev_process_count_dag
)


**Output (After executing the DAG):**
AIRFLOW_CTX_DAG_OWNER=airflow
AIRFLOW_CTX_DAG_ID=aws_dev_process_count_dag
AIRFLOW_CTX_TASK_ID=jps_process_count
AIRFLOW_CTX_EXECUTION_DATE=2021-11-01T08:30:36.500479+00:00
AIRFLOW_CTX_DAG_RUN_ID=manual__2021-11-01T08:30:36.500479+00:00
[2021-11-01 04:30:40,085] {ssh.py:203} WARNING - Remote Identification Change is not verified. This wont protect against Man-In-The-Middle attacks
[2021-11-01 04:30:40,089] {ssh.py:210} WARNING - No Host Key Verification. This wont protect against Man-In-The-Middle attacks
[2021-11-01 04:30:40,121] {transport.py:1819} INFO - Connected (version 2.0, client OpenSSH_8.0)
[2021-11-01 04:30:40,279] {transport.py:1819} INFO - Authentication (publickey) successful!
[2021-11-01 04:30:40,279] {ssh.py:110} INFO - Running command: 
   set -e;
   ps_count=`jps | grep -v Jps | wc -l`
   if [ $ps_count -ne 0 ]
   then
       echo "ps_count is $ps_count, There might be some processes are running"
       exit 1
   else
       echo "All processed were stopped..!"
   fi
[2021-11-01 04:30:40,757] {ssh.py:142} INFO - ps_count is 38, There might be some processes are running
[2021-11-01 04:30:40,763] {taskinstance.py:1455} ERROR - SSH operator error: error running cmd: 
   set -e;
   ps_count=`jps | grep -v Jps | wc -l`
   if [ $ps_count -ne 0 ]
   then
       echo "ps_count is $ps_count, There might be some processes are running"
       exit 1
   else
       echo "All processed were stopped..!"
   fi, error: 
Traceback (most recent call last):
  File "/home/transcenddev7/.local/lib/python3.8/site-packages/airflow/providers/ssh/operators/ssh.py", line 169, in execute
    raise AirflowException(f"error running cmd: {self.command}, error: {error_msg}")
airflow.exceptions.AirflowException: error running cmd: 
   set -e;
   ps_count=`jps | grep -v Jps | wc -l`
   if [ $ps_count -ne 0 ]
   then
       echo "ps_count is $ps_count, There might be some processes are running"
       exit 1
   else
       echo "All processed were stopped..!"
   fi, error: 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/transcenddev7/.local/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 1112, in _run_raw_task
    self._prepare_and_execute_task_with_callbacks(context, task)
  File "/home/transcenddev7/.local/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 1285, in _prepare_and_execute_task_with_callbacks
    result = self._execute_task(context, task_copy)
  File "/home/transcenddev7/.local/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 1315, in _execute_task
    result = task_copy.execute(context=context)
  File "/home/transcenddev7/.local/lib/python3.8/site-packages/airflow/providers/ssh/operators/ssh.py", line 172, in execute
    raise AirflowException(f"SSH operator error: {str(e)}")
airflow.exceptions.AirflowException: SSH operator error: error running cmd: 
   set -e;
   ps_count=`jps | grep -v Jps | wc -l`
   if [ $ps_count -ne 0 ]
   then
       echo "ps_count is $ps_count, There might be some processes are running"
       exit 1
   else
       echo "All processed were stopped..!"
   fi, error: 
[2021-11-01 04:30:40,765] {taskinstance.py:1496} INFO - Marking task as FAILED. dag_id=aws_dev_process_count_dag, task_id=jps_process_count, execution_date=20211101T083036, start_date=20211101T083039, end_date=20211101T083040
[2021-11-01 04:30:41,235] {local_task_job.py:146} INFO - Task exited with return code 1

**Reply by potiuk** [reply]
Well. From what I see this is what you get - fail scenario.

Your command exits with return code 1. and AirflowException is raised, which causes the task to fail:

| Marking task as FAILED

 What else did you expect to happen ?

**Reply by dgedela** [reply]
Thank you potuik..!

I received mail on-failure like below. (It is saying that it is SSH operator error)

Try 1 out of 1
Exception:
**SSH operator error: error running cmd:** 
set -e;
ps_count=`jps | grep -v Jps | wc -l`
if [ $ps_count -ne 0 ]
then
echo "ps_count is $ps_count, There might be some processes are running"
exit 1
else
echo "All processed were stopped..!"
fi, error: 


**what we expected:**
If it is graceful error then it should only throw like "ps_count is non-zero, There might be some processes are running..! So that, it was failed" but not SSH Operator error.

**Reply by potiuk** [reply]
Ah. NOW you got to the real problem that you think you have. It's always good to write what your "expectation" is and seems that finally you expressed that in the 4th iteration. It's the first time you write that you have a problem with  email and it's content.

The problem is that you should print your error to `stderr` if you want to see the output. Something like that should do:
```
>&2 echo "ps_count is $ps_count, There might be some processes are running"
```
